virtcontainers/ich_arch_base.go

```diff
@@ -0,0 +1,278 @@
+package virtcontainers
+
+
+import (
+	
+	"fmt"
+	"strings"
+
+	"github.com/kata-containers/runtime/virtcontainers/types"
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	
+	govmmIch "github.com/kata-containers/runtime/virtcontainers/pkg/intel/govmm/ich"
+
+)
+
+const (
+	defaultIchPath            = "/usr/local/bin/cloud-hypervisor"
+	defaultIchMachineType     = "x86_64"
+)
+
+type ichArch interface {
+	
+	// ichPath returns the path to the Ich binary
+	ichPath() (string, error)
+
+	// kernelParameters returns the kernel parameters
+	kernelParameters(debug bool) []Param
+
+	// appendImage appends an image to devices
+	appendImage(devices []govmmIch.Device, path string) ([]govmmIch.Device, error)
+
+	// appendNetwork appends a endpoint device to devices
+	appendNetwork(devices []govmmIch.Device, endpoint Endpoint) ([]govmmIch.Device, error)
+
+	// appendHybridVSock appends a hybrid vsock PCI to devices
+	appendHybridVSock(devices []govmmIch.Device, vsock types.HybridVSock) ([]govmmIch.Device, error)
+	
+	appendRNGDevice(devices []govmmIch.Device, rngDev config.RNGDev) ([]govmmIch.Device, error)
+	
+	appendSerialConsole(devices []govmmIch.Device, consoletype string) ([]govmmIch.Device, error)
+	
+	appendVirtioConsole(devices []govmmIch.Device, consoletype string) ([]govmmIch.Device, error)
+	
+	appendApiSocket(devices []govmmIch.Device, apiSocket string) ([]govmmIch.Device, error)
+	
+	appendVirtualFilesystem(devices []govmmIch.Device, tag string, path string) ([]govmmIch.Device, error)
+	
+	appendMemory(memorySize uint64, path string) (govmmIch.Memory, error)
+	
+	appendProcessors(vcpu uint32) (govmmIch.VCPU, error)
+
+	//capabilities returns the capabilities supported by ICH
+	capabilities() types.Capabilities
+	
+}
```

Comment:
Inline with your previous comment I can remove this but bear in mind that the main reason I added it was not the abstraction but the state aggregation into a device list before generating the final command line argument. This mechanism has to stay in the driver somehow because this builds by the 'add' calls into the driver (nic's are being added late in the process for instance) after which the state is ready to be translated into a working hypervisor. So to cut it short, I still need some state aggregation in the driver for devices if we cut off the iovmm logic.

How do you see this?

Reply:
Oh yes, you're going to implement a way to aggregate devices and other parameters into a command line string for now (an HTTP payload later on). What I'm saying is that it should be part of the driver itself, not a govmm implementation.

Reply:
Very good, then we're on the same page. I'll fix it.

Reply:
Done.